zooniverse/Panoptes-Front-End
# Pages Editor: UI tweaks, improved Workflow Settings

## PR Overview

Part of: **Pages Editor MVP** project and **FEM Lab** super-project
Follows #7088
Staging branch URL: https://pr-7093.pfe-preview.zooniverse.org/lab/1982/workflows/editor/3711?env=staging

This PR combines a number of UI tweaks and updates to the Workflow Settings. The UI changes makes the visual presentation closer to [Sean's intended design](https://www.figma.com/file/gNOcFdlskfEXmVIFPvx7Yj/Pages-Editor?type=design&node-id=0-1&mode=design), while the Workflow Settings changes make that tab ⭐ _on par with the Project Builder FEM Lab's functionality._ [(comparative link)](https://master.pfe-preview.zooniverse.org/lab/1982/workflows/3711?femLab=true&env=staging)

**Part 1: UI changes**
- Workflow ID moved to Workflow Settings tab
- **Workflow Version** added, and also displays a spinner 🕐 to indicate loading/fetching activity
- TasksPage: now displays a "you have no tasks" notice if you have no tasks.

<img width="300" alt="image" src="https://github.com/zooniverse/Panoptes-Front-End/assets/13952701/bd29d2ca-6246-4998-b7a2-be45d328951a"> ➡️  <img width="300" alt="image" src="https://github.com/zooniverse/Panoptes-Front-End/assets/13952701/8c00f4e3-c907-43fe-a439-1956141aeaa1">

<img width="300" alt="image" src="https://github.com/zooniverse/Panoptes-Front-End/assets/13952701/0fe17209-fa22-4da4-999d-53d4aa49ea88"> ➡️ <img width="300" alt="image" src="https://github.com/zooniverse/Panoptes-Front-End/assets/13952701/dcca584c-001b-4bef-8cd8-0e7dfb8fbce6">

<img width="300" alt="image" src="https://github.com/zooniverse/Panoptes-Front-End/assets/13952701/8e60a66b-01ee-4d07-ac86-a0cd00821c3f">

**Part 2: Workflow Settings Tab**
- Subject Retirement rule now locked to Classification Count, with no ability to specify _how many_ to retirement.
- Multi Image Options rules added
  - Enable "Allow Separate Frames View" to show additional options.
- Image Display Options now includes "Allow users to flip color"
- Hidden features:
  - you can now access the Workflow Settings tab directly by adding `?tab=settings` to the URL
  - you can now access "removed options" (i.e. workflow options that are valid in the API but removed to keep things in line with Project Builder FEM Lab) by adding ~?tab=settings&showRemovedOptions=true~ `?tab=settings&advanced=true` to the URL

**Misc Changes**
- ❗ DataManager: the data memo is now updated when the ready/fetching/updating `status` value changes. This wasn't tracked previously!
  - Since I'm updating the data manager, I need to pay extra attention that that changes are being committed to the API properly, and updates from the API are being fetched properly. **Reviewer: I'd appreciate eyes on this as well.**
- The Experimental/Debug Panel's "Reset" button now also resets `workflow.configuration`
- Icons added: WrenchIcon, LoadingIcon (spinner)

**Update 1** [(link)](https://github.com/zooniverse/Panoptes-Front-End/pull/7093#issuecomment-2097156613)
- ⭐ Workflows are now **linear.**
- New advanced mode hidden behind `?advanced=true`. (Use this to enable manual/non-linear workflows, and the Experimental Panel)
- Fix: classifications count no longer hidden

### Testing

No specific tests recommended. Just check that general functionality is maintained, and the code makes sense.

- Open the Test URL
- Make general changes to the workflow on the Tasks tab:
  - If no tasks, "you have no tasks" notification should appear
  - You should be able to add tasks/pages, edit them, etc 
- Check the Workflow Settings Tab:
  - You should be able to change workflow-level settings, e.g. change the workflow's name.
  - The amount of workflow-level settings you can change should match the FEM Lab (?femLab=true) version of the Project Builder: https://master.pfe-preview.zooniverse.org/lab/1982/workflows/3711?femLab=true&env=staging
- Reload the page to confirm changes are made.

### Status

Ready for review. 👌 

Requires 7088 to be merged first before this can be merged.
